File: Week_01/java/src/_206_ReverseLinkedList.java
/**
 * 第一遍：2020/08/14周六 ✅
 * 第二遍：2020/08/14周六
 * 第三遍：2020/05/25周一
 * 第四遍：2020/08/13周四
 * 第五遍：2020/08/14周五
 */
class _206_ReverseLinkedList {
    //时间O(n) 空间O(n)
    public ListNode reverseList(ListNode head) {
        if (head == null) return null;
        if (head.next == null) return head;
        ListNode headNew = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return headNew;
    }

    class SolutionIterative {
        //时间O(n) 空间O(1)
        public ListNode reverseList(ListNode head) {
            if (head == null) return null;
            ListNode headReversed = null;
            while (head != null) {
                ListNode next = head.next;
                head.next = headReversed;
                headReversed = head;
                head = next;
            }
            return headReversed;
        }
    }

    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }
}

File: Week_01/kotlin/src/_206_ReverseLinkedList.kt
fun reverseList(head: ListNode?): ListNode? {
    var headNew: ListNode? = null
    var node = head
    while (node != null) {
        val curr = node
        node = node.next
        curr.next = headNew
        headNew = curr
    }
    return headNew
}

class _206_Recursive_ {
    fun reverseList(head: ListNode?): ListNode? {
        if (head?.next == null) return head
        val tail = head.next
        val headNew = reverseList(tail)
        tail?.next = head
        head.next = null
        return headNew
    }
}